---
title: "Differential expression"
output: html_document
---
```{r setup, include=FALSE}
source("functions.R")
library(org.Hs.eg.db)
library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(edgeR)
library(DT)
sce.pass <- readRDS("results/sce_pass.rds")
sce.pass_hcc <- readRDS("results/sce_pass_hcc_pseudo.rds")
```

# Wilcoxon test for DE

I have already tested for previous analysis on other single cell projects various different types of differential expression tests and have found a simple wilcoxon test to work the best for those data. I will therefore start with the differential expression for this data using the wilcoxon test.

# Set up gene lengths for RPKM

```{r, eval=FALSE}
GTFfile = "refcoding.gtf.gz"

#Load the annotation and reduce it
GTF <- import.gff(GTFfile, format="gtf", genome="GRCh37.75", feature.type="exon")
grl <- GenomicRanges::reduce(split(GTF, elementMetadata(GTF)$gene_id))
reducedGTF <- unlist(grl, use.names=T)
elementMetadata(reducedGTF)$gene_id <- rep(names(grl), lengths(grl))

elementMetadata(reducedGTF)$widths <- width(reducedGTF)

#Create a list of the ensembl_id/GC
calc_GC_length <- function(x) {
  sum(elementMetadata(x)$widths)
}
output <- sapply(split(reducedGTF, elementMetadata(reducedGTF)$gene_id), calc_GC_length)

output <- as.data.frame(output)
```

# Perform DE on Hypoxia vs Normoxia MCF7

```{r, eval=FALSE}
groupA.m <- exprs(sce.pass[,as.matrix(colData(sce.pass)["Condition"] %in% "Norm")])
groupA.m <- na.omit(groupA.m)
groupB.m <- exprs(sce.pass[,as.matrix(colData(sce.pass)["Condition"] %in% "Hypo")])
groupB.m <- na.omit(groupB.m)


m.A <- match(rownames(groupA.m), rownames(output))
gene.lengths <- output$output[m.A]

RPKM.A <- rpkm(groupA.m, gene.length = gene.lengths,normalized.lib.sizes=TRUE)

m.B <- match(rownames(groupB.m), rownames(output))
gene.lengths <- output$output[m.B]

RPKM.B <- rpkm(groupB.m, gene.length = gene.lengths,normalized.lib.sizes=TRUE)



fishersMethod <-function(x) pchisq(-2 * sum(log(x)),df=2*length(x),lower.tail=FALSE)

# set up empty data frame
result.f<-data.frame()

# loop over nrow groupA.m i.e. for each cell
for(i in 1:nrow(RPKM.A)){

n<-RPKM.A[i,] # take the i'th sample from groupA
m<-RPKM.B[i,] # take the i'th sample from groupB
n.f<-n[n>0] # subset all data for genes with count >0 in groupA
m.f<-m[m>0] # subset all data for genes with count >0 in groupB

x<-length(n) # calulate length of groupA
y<-length(m) # calculate length of groupB

x.1<-length(n.f) # calculate length of n.f groupA
y.1<-length(m.f) # calulate length of m.f of groupB
x.2<-x-x.1 # calulate difference between expression in groupA
y.2<-y-y.1 # calulate difference between expression in groupB

m.m <- matrix(c(x.1, x.2, y.1, y.2), ncol = 2) # now set up a matrix (contingency table) of sample expression A vs B
fisher.test.p <-fisher.test(m.m)$p.value # Now use fishers to test if the comparrison between cell frequency is different in group A vs groupB

wilcox.p<-wilcox.test(n,m,paired = FALSE)$p.value # wilcoxon test for difference in expression. ?? However, shouldnt it now be done on genes not rows? Am I missing something? Clarify with supat?

all.p<-c(fisher.test.p,wilcox.p) # combined test
fisher.p<-fishersMethod (all.p) # Then perform a fishers method function

my.meanA<-mean(RPKM.A[i,]) # calculate the mean of groupA
my.meanB<-mean(RPKM.B[i,]) # calulcate the mean of groupB
my.log2.fc<-log2(my.meanB)-log2(my.meanA) # calulate the difference in mean
my.test.f<-data.frame(gene=rownames(RPKM.A)[i],meanA=my.meanA,meanB=my.meanB,
log2fc=my.log2.fc,nCellA=x,nCellB=y,expCellA=x.1,expCellB=y.1,
expFractionCellA=x.1/x,expFractionCellB=y.1/y,fisher.test.p=fisher.test.p,wilcox=wilcox.p,fisher=fisher.p) # make a huge dataframe will all of the analysis in
result.f<-rbind(result.f,my.test.f) # Then bind the result to my tests dataframe.
}

result.f<-unique(result.f[order(result.f$wilcox),]) # order the result by fishers test
result.f$p.adjust<-p.adjust(result.f$fisher, method ="BH") # Then perform benjamini-hochberg adjustment

result.f <- ensembl_to_symbol(result.f, "gene")

write.table(result.f,file="results/HypovsNorm.txt", append=FALSE, sep="\t", quote=FALSE,row.names=FALSE, col.names=TRUE)
```

# output gene table

The following table contains the data for all genes normoxia vs hypoxia

```{r, echo=FALSE, message=FALSE, warning=FALSE}
result.f <- read.csv("results/HypovsNorm.txt", sep="\t")

# Filter mean A and B columns
result.f <- result.f %>% 
  filter(meanA > 1 & meanB > 1)

datatable(result.f)
```

# Upregulated genes

The following table contains the differential gene expression for the upregulated genes for normoxia vs hypoxia

```{r, echo=FALSE, message=FALSE, warning=FALSE}
result.f <- read.csv("results/HypovsNorm.txt", sep="\t")

# Filter mean A and B columns
result.f <- result.f %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc > 1) %>% 
  arrange(desc(log2fc))

datatable(result.f)
```

# Downregulated genes

The following table contains the differential gene expression for the downregulated genes for normoxia vs hypoxia

```{r, echo=FALSE, message=FALSE, warning=FALSE}
result.f <- read.csv("results/HypovsNorm.txt", sep="\t")

# Filter mean A and B columns
result.f <- result.f %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc < -1) %>% 
  arrange(log2fc)

datatable(result.f)
```






# Perform DE on Hypoxia vs Normoxia HCC

```{r, eval=FALSE}
groupA.m <- exprs(sce.pass_hcc[,as.matrix(colData(sce.pass_hcc)["Condition"] %in% "Normoxia")])
groupA.m <- na.omit(groupA.m)
groupB.m <- exprs(sce.pass_hcc[,as.matrix(colData(sce.pass_hcc)["Condition"] %in% "Hypoxia")])
groupB.m <- na.omit(groupB.m)


m.A <- match(rownames(groupA.m), rownames(output))
gene.lengths <- output$output[m.A]

RPKM.A <- na.omit(rpkm(groupA.m, gene.length = gene.lengths,normalized.lib.sizes=TRUE))

m.B <- match(rownames(groupB.m), rownames(output))
gene.lengths <- output$output[m.B]

RPKM.B <- na.omit(rpkm(groupB.m, gene.length = gene.lengths,normalized.lib.sizes=TRUE))



fishersMethod <-function(x) pchisq(-2 * sum(log(x)),df=2*length(x),lower.tail=FALSE)

# set up empty data frame
result.f<-data.frame()

# loop over nrow groupA.m i.e. for each cell
for(i in 1:nrow(RPKM.A)){

n<-RPKM.A[i,] # take the i'th sample from groupA
m<-RPKM.B[i,] # take the i'th sample from groupB
n.f<-n[n>0] # subset all data for genes with count >0 in groupA
m.f<-m[m>0] # subset all data for genes with count >0 in groupB

x<-length(n) # calulate length of groupA
y<-length(m) # calculate length of groupB

x.1<-length(n.f) # calculate length of n.f groupA
y.1<-length(m.f) # calulate length of m.f of groupB
x.2<-x-x.1 # calulate difference between expression in groupA
y.2<-y-y.1 # calulate difference between expression in groupB

m.m <- matrix(c(x.1, x.2, y.1, y.2), ncol = 2) # now set up a matrix (contingency table) of sample expression A vs B
fisher.test.p <-fisher.test(m.m)$p.value # Now use fishers to test if the comparrison between cell frequency is different in group A vs groupB

wilcox.p<-wilcox.test(as.numeric(n),as.numeric(m),paired = FALSE)$p.value # wilcoxon test for difference in expression. ?? However, shouldnt it now be done on genes not rows? Am I missing something? Clarify with supat?

all.p<-c(fisher.test.p,wilcox.p) # combined test
fisher.p<-fishersMethod (all.p) # Then perform a fishers method function

my.meanA<-mean(RPKM.A[i,]) # calculate the mean of groupA
my.meanB<-mean(RPKM.B[i,]) # calulcate the mean of groupB
my.log2.fc<-log2(my.meanB)-log2(my.meanA) # calulate the difference in mean
my.test.f<-data.frame(gene=rownames(RPKM.A)[i],meanA=my.meanA,meanB=my.meanB,
log2fc=my.log2.fc,nCellA=x,nCellB=y,expCellA=x.1,expCellB=y.1,
expFractionCellA=x.1/x,expFractionCellB=y.1/y,fisher.test.p=fisher.test.p,wilcox=wilcox.p,fisher=fisher.p) # make a huge dataframe will all of the analysis in
result.f<-rbind(result.f,my.test.f) # Then bind the result to my tests dataframe.
}

result.f<-unique(result.f[order(result.f$wilcox),]) # order the result by fishers test
result.f$p.adjust<-p.adjust(result.f$fisher, method ="BH") # Then perform benjamini-hochberg adjustment

result.f <- ensembl_to_symbol(result.f, "gene")

write.table(result.f,file="results/HypovsNorm_hcc.txt", append=FALSE, sep="\t", quote=FALSE,row.names=FALSE, col.names=TRUE)
```

# output gene table

The following table contains the data for all genes normoxia vs hypoxia

```{r, echo=FALSE, message=FALSE, warning=FALSE}
result.f <- read.csv("results/HypovsNorm_hcc.txt", sep="\t")

# Filter mean A and B columns
result.f <- result.f %>% 
  filter(meanA > 1 & meanB > 1)

datatable(result.f)
```

# Upregulated genes

The following table contains the differential gene expression for the upregulated genes for normoxia vs hypoxia

```{r, echo=FALSE, message=FALSE, warning=FALSE}
result.f <- read.csv("results/HypovsNorm_hcc.txt", sep="\t")

# Filter mean A and B columns
result.f <- result.f %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc > 1) %>% 
  arrange(desc(log2fc))

datatable(result.f)
```

# Downregulated genes

The following table contains the differential gene expression for the downregulated genes for normoxia vs hypoxia

```{r, echo=FALSE, message=FALSE, warning=FALSE}
result.f <- read.csv("results/HypovsNorm_hcc.txt", sep="\t")

# Filter mean A and B columns
result.f <- result.f %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc < -1) %>% 
  arrange(log2fc)

datatable(result.f)
```


# Comparrison of overlaps etween MCF7 and HCC

```{r}
mcf7 <- read.csv("results/HypovsNorm.txt", sep="\t")
hcc <- read.csv("results/HypovsNorm_hcc.txt", sep="\t")


mcf7_down <- mcf7 %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc < -1) %>% 
  arrange(log2fc)

mcf7_up <- mcf7 %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc > 1) %>% 
  arrange(log2fc)

hcc_down <- hcc %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc < -1) %>% 
  arrange(log2fc)

hcc_up <- hcc %>% 
  filter(meanA > 1 & meanB > 1) %>% 
  filter(log2fc > 1) %>% 
  arrange(log2fc)


genes_down <- intersect(mcf7_down$SYMBOL, hcc_down$SYMBOL)
genes_up <- intersect(mcf7_up$SYMBOL, hcc_up$SYMBOL)


genes <- c(genes_down, genes_up)
```

# Heatmap of overlaps

```{r}
sce.pass_hcc <- readRDS("results/sce_pass_hcc.rds")
sce.pass_mcf7 <- readRDS("results/sce_pass.rds")
colData(sce.pass_hcc)["Cell"] <- "HCC"

Condition <- colData(sce.pass_hcc)["Condition"]

Condition <- as.data.frame(Condition) 
Condition <- Condition %>% 
  mutate(Condition = replace(Condition, Condition == "Hypoxia", "Hypo"))
Condition <- Condition %>% 
  mutate(Condition = replace(Condition, Condition == "Normoxia", "Norm"))

colData(sce.pass_hcc)["Condition"] <- Condition$Condition

colData(sce.pass_mcf7)["Cell"] <- "MCF"  

sce.pass <- cbind(logcounts(sce.pass_hcc), logcounts(sce.pass_mcf7))
sce.pass.coldata <- rbind(colData(sce.pass_hcc)[c("Condition", "Cell", "assignments")], colData(sce.pass_mcf7)[c("Condition", "Cell", "assignments")])
library(pheatmap)

ann <- as.data.frame(sce.pass.coldata)

matrix <- sce.pass
matrix <- matrix - rowMeans(matrix)
matrix <- subset(matrix, rownames(matrix) %in% genes)

svg("Figures.dir/Grouped_diff_exprs.svg")
pheatmap(matrix, fontsize=5, annotation = ann)
dev.off()
```

