---
title: "Combined mixed species seurat analysis"
---

This Markdown is contructed to analyse a combination of singlce cell experiments that are composed of mixed species.

My main reason for using Seurat over other single-cell software is that it contains an implimentation of Cononical Correlation Alignment (CCA) that allows different datasets to be merged.


# Import the packages

```{r, include=FALSE}
knitr::opts_chunk$set(cache=FALSE, warning = FALSE, message = FALSE)
Sys.setenv(KMP_DUPLICATE_LIB_OK=TRUE)
library(Seurat)
library(tximport)
library(cowplot)
library(org.Hs.eg.db)
library(dplyr)
library("AnnotationDbi")
library(EnsDb.Hsapiens.v75)
library(EnsDb.Mmusculus.v75)
library(stringr)
meta_data <- readr::read_csv("meta_data.csv")
sample_files <- meta_data$Sample

```

# Import the data 

The experiment was analysed using [cgat-developer](https://github.com/cgat-developers/cgat-core) pipelines for quality of sequencing. Read quality looked good. 

Next I quanitified the experssion using [Slamon Alevin method](https://salmon.readthedocs.io/en/latest/alevin.html).  

```{r}
for (i in sample_files){
  txi.import <- tximport(gsub("SAMPLE_FILE",i , "salmon.dir/SAMPLE_FILE/alevin/quants_mat.gz"), type="alevin")
  d <- txi.import$counts
  
  SYMBOL <- mapIds(EnsDb.Hsapiens.v75,keys=rownames(d),column="SYMBOL",keytype="GENEID",
                   multiVals="first")
  rownames(d) <- make.unique(ifelse(is.na(SYMBOL), rownames(d), SYMBOL))
    SYMBOL <- mapIds(EnsDb.Mmusculus.v75,keys=rownames(d),column="SYMBOL",keytype="GENEID",
                   multiVals="first")
  rownames(d) <- make.unique(ifelse(is.na(SYMBOL), rownames(d), SYMBOL))

  seurat_obj <- CreateSeuratObject(counts = d, min.cells = 3, min.features = 200, project = "pbmc1")
  seurat_obj$cells <- meta_data[meta_data$Sample == i,]$treatment
  assign(paste("seurat_obj", i, sep = "."), seurat_obj)
  
}

```

# Assess QC metrics {.tabset .tabset-fade}


# Vln plots of features {.tabset}

In the example below, we visualize QC metrics, and use these to filter cells.

We filter cells that have unique feature counts over 2,500 or less than 200
We filter cells that have >5% mitochondrial counts

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seurat <- get(gsub("SAMPLE_FILE",i , "seurat_obj.SAMPLE_FILE"))
  seurat[["percent.mt"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
cat("## ",i,"\n")
plt <- VlnPlot(seurat, c("nFeature_RNA", "nCount_RNA", "percent.mt"))
print(plt)
cat('\n\n')
  assign(paste("seurat_obj", i, sep = "."), seurat)
}
```

# Feature plots {.tabset}

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seurat <- get(gsub("SAMPLE_FILE",i , "seurat_obj.SAMPLE_FILE"))
  plot1 <- FeatureScatter(seurat, feature1 = "nCount_RNA", feature2 = "percent.mt")
  plot2 <- FeatureScatter(seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
  cat("## ",i,"\n")
  plot(CombinePlots(plots = list(plot1, plot2)))
  cat('\n\n')
}
```


# Filter the cells

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seurat <- get(gsub("SAMPLE_FILE",i , "seurat_obj.SAMPLE_FILE"))
  seurat <- subset(seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 25)
  assign(paste("seurat_obj", i, sep = "."), seurat)
}

```

# Normalise the data

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seurat <- get(gsub("SAMPLE_FILE",i , "seurat_obj.SAMPLE_FILE"))
  seurat <- NormalizeData(seurat)
  assign(paste("seurat_obj", i, sep = "."), seurat)
}
```

# Identify highly variable genes {.tabset}

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

The procedure in Seurat3 is described in detail here, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the FindVariableFeatures function. By default, we return 2,000 features per dataset. These will be used in downstream analysis, like PCA.

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seurat <- get(gsub("SAMPLE_FILE",i , "seurat_obj.SAMPLE_FILE"))
  seurat <- FindVariableFeatures(seurat, selection.method = "disp", nfeatures = 2000)
  # Identify the 10 most highly variable genes
  top10 <- head(VariableFeatures(seurat), 10)

  #plot variable features with and without labels
  plot1 <- VariableFeaturePlot(seurat)
  plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
  cat("## ",i,"\n")
  print(plot2)
  assign(paste("seurat_obj", i, sep = "."), seurat)
  cat('\n\n')
}
```

# Integrate both datasets 

Next I will use Seurat to integrate the datasets together into one object so I can compare clusters between the two datasets.

```{r}
meta_data <- readr::read_csv("meta_data.csv")
sample_files <- meta_data$Sample
samples <- mget(paste0("seurat_obj.",sample_files))
integration <- FindIntegrationAnchors(object.list = samples, 
                                      dims=1:10,
                                      k.filter = 150)
to_integrate <- rownames(x = integration)
integration <- IntegrateData(anchorset = integration, dims = 1:10, features.to.integrate = to_integrate)

```

# Evaluate the clustering

Next I will use UMAP to visualise the clustering

First, we need to apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData function:

Shifts the expression of each gene, so that the mean expression across cells is 0
Scales the expression of each gene, so that the variance across cells is 1
This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
The results of this are stored in integration[["RNA"]]@scale.data

```{r}
DefaultAssay(integration) <- "integrated"
integration <- ScaleData(integration)
```

# PCA

```{r}
integration <- RunPCA(integration, npcs = 30)
DimPlot(integration, reduction = "pca", split.by = "cells")
```

# UMAP

```{r}
integration <- RunUMAP(integration, reduction = "pca", dims = 1:20)
integration <- FindNeighbors(integration, reduction = "pca", dims = 1:20)
integration <- FindClusters(integration, resolution = 0.4)

p1 <- DimPlot(integration, reduction = "umap", group.by = "cells")
p2 <- DimPlot(integration, reduction = "umap", label = TRUE)
plot_grid(p1,p2)
```

# Clustering facet

```{r}
DimPlot(integration, reduction = "umap", split.by = "cells")

DimPlot(integration, reduction = "umap")
```


# Measure human and mouse reads {.tabset}

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  txi.import <- tximport(gsub("SAMPLE_FILE",i , "salmon.dir/SAMPLE_FILE/alevin/quants_mat.gz"), type="alevin")
  d <- txi.import$counts
  gene_species <- ifelse(str_detect(rownames(d), "^ENSMUSG"), "mouse", "human")
  mouse_inds <- gene_species == "mouse"
  human_inds <- gene_species == "human"
  cell_species <- tibble(n_mouse_umi = Matrix::colSums(d[mouse_inds,]),
                       n_human_umi = Matrix::colSums(d[human_inds,]),
                       tot_umi = Matrix::colSums(d),
                       prop_mouse = n_mouse_umi / tot_umi,
                       prop_human = n_human_umi / tot_umi)
  cell_species <- cell_species %>% 
  mutate(species = case_when(
    prop_mouse > 0.9 ~ "mouse",
    prop_human > 0.9 ~ "human",
    TRUE ~ "mixed"
  ))
  gg <- ggplot(cell_species, aes(n_human_umi, n_mouse_umi, color = species)) +
  geom_point(size = 0.5)
  
  assign(paste("cell_species", i, sep = "."), cell_species)
  assign(paste("d", i, sep = "."), d)
  cat("## ",i,"\n")
  print(gg)
  cat('\n\n')

  
}
```


# Number of cells per species {.tabset}

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  cell_species <- get(paste("cell_species", i, sep = "."))
  t <- cell_species %>% 
  dplyr::count(species) %>% 
  mutate(proportion = n / ncol(d))

  cat("## ",i,"\n")
  kableExtra::kable(t)
  cat('\n\n')
} 
```


# Vln plot of each cells {.tabset}

```{r, results='asis', echo = FALSE}

for (i in sample_files){
  d <- get(gsub("SAMPLE_FILE",i , "d.SAMPLE_FILE"))
  cell_species <- get(gsub("SAMPLE_FILE",i , "cell_species.SAMPLE_FILE"))
  seu <- CreateSeuratObject(d, min.cells = 3, min.features = 200) %>% 
  NormalizeData(verbose = FALSE) %>% 
  ScaleData(verbose = FALSE) %>% 
  FindVariableFeatures(verbose = FALSE, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 25)
  seu <- AddMetaData(seu, metadata = cell_species$species, col.name = "species")
  vln <- VlnPlot(seu, c("nCount_RNA", "nFeature_RNA"), group.by = "species",
        pt.size = 0.1)
  seu <- RunPCA(seu, verbose = FALSE, npcs = 30)
  assign(paste("seu", i, sep = "."), seu)
  cat("## ",i,"\n")
  print(vln)
  cat('\n\n')
  }

```

# PCA plot of samples {.tabset}

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seu <- get(paste0("seu.", as.character(i)))
  plt <- DimPlot(seu, reduction = "pca", pt.size = 0.5, group.by = "species")

  cat("## ",i,"\n")
  print(plt)
  cat('\n\n')
  }
 
```

# UMAP plots {.tabset}

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  seu <- get(paste0("seu.", as.character(i)))
  seu <- RunUMAP(seu, reduction = "pca", dims = 1:20)
  umap <- DimPlot(seu, reduction = "umap", pt.size = 0.5, group.by = "species")

  cat("## ",i,"\n")
  print(umap)
  cat('\n\n')
}
```



# Integrate both datasets 

Next I will use Seurat to integrate the datasets together into one object so I can compare clusters between the two datasets.

```{r}
meta_data <- readr::read_csv("meta_data.csv")
sample_files <- meta_data$Sample
samples <- mget(paste0("seu.",sample_files))
integration <- FindIntegrationAnchors(object.list = samples, 
                                      dims=1:10,
                                      k.filter = 150)
to_integrate <- rownames(x = integration)
integration <- IntegrateData(anchorset = integration, dims = 1:10, features.to.integrate = to_integrate)

```

# Evaluate the clustering

Next I will use UMAP to visualise the clustering

First, we need to apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData function:

Shifts the expression of each gene, so that the mean expression across cells is 0
Scales the expression of each gene, so that the variance across cells is 1
This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
The results of this are stored in integration[["RNA"]]@scale.data

```{r}
DefaultAssay(integration) <- "integrated"
integration <- ScaleData(integration)
```

# PCA {.tabset}

```{r}
integration <- RunPCA(integration, npcs = 30)
plt <- DimPlot(integration, reduction = "pca", split.by = "species")
  cat("## ",i,"\n")
  print(plt)
  cat('\n\n')
```

# UMAP {.tabset}

```{r}
integration <- RunUMAP(integration, reduction = "pca", dims = 1:10)
integration <- FindNeighbors(integration, reduction = "pca", dims = 1:10)
integration <- FindClusters(integration, resolution = 0.4)

p1 <- DimPlot(integration, reduction = "umap", group.by = "species")
p2 <- DimPlot(integration, reduction = "umap", label = TRUE)


cat("## ",'By species',"\n")
print(p1)
  cat('\n\n')


cat("## ",'By cluster',"\n")
print(p1)
cat('\n\n')

```
